### PR TITLE
Generalize Configuration Functions Spec.

### PR DESCRIPTION
**Pull request by frankfarzan**
Configuration functions as defined functions-spec.md
can be implemented using any toolchain and invoked using any container
workflow orchestrator like Tekton, Cloud Build, or run directly using `docker run`.

functions-impl describes using `kustomize config run` as an example orchestrator
for invoking configuration functions.

### CODE REVIEW

**Comment by pwittrock** [comment]
Lets defer deleting the existing document until we know how we want to replace it in the command.

I think we want to publish all documentation built into the cli itself so it is available and discoverable.

File: cmd/config/docs/api-conventions/functions-spec.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
+# Configuration Functions Specification
+
+This document specifies a standard for client-side functions that operate on Kubernetes declarative configurations. This standard enables creating small, interoperable, and language-independent executable programs packaged as containers that can be chained together as part of a configuration management pipeline. The end result of such a pipeline are fully rendered configurations that can then be applied to a control plane (e.g. Using ‘kubectl apply’ for Kubernetes control plane). As such, although this document references Kubernetes Resource Model and API conventions, it is completely decoupled from Kuberentes API machinery and does not depend on any in-cluster components.
+
+This document references terms described in [Kubernetes API Conventions][1].
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119][2].
+
+## Applications
```
Comment (pwittrock) [inline cmd/config/docs/api-conventions/functions-spec.md]: Use Cases?  Motivation?

Applications is confusing as a term -- e.g. App as in workload or App as in use case

File: cmd/config/docs/api-conventions/functions-spec.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  Type, annotations, etc.
+
+Performing these on the client rather than the server enables:
+
+- Configuration to be reviewed prior to being sent to the API server
+- Configuration to be validated as part of the CD pipeline
+- Configuration for Resources to validated holistically rather than individually
+  per-Resource -- e.g. ensure the `Service.selector` and `Deployment.spec.template` labels
+  match.
+- MutatingWebHooks are scoped to a single Resource instance at a time.
```
Comment (pwittrock) [inline cmd/config/docs/api-conventions/functions-spec.md]: this should probably be indented

File: cmd/config/docs/api-conventions/functions-spec.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+### Serialization Format
+
+A function MUST support YAML as a serialization format for the input and output. A function MUST use utf8 encoding (as YAML is a superset of JSON, JSON will also be supported by any conforming function).
+
+### Operations
+
+A function MAY Create, Update, or Delete any number of items in the `items` field and output the resultant list.
+
+A function MAY modify annotations with prefix `config.kubernetes.io`, but must be careful about doing so since they’re used for orchestration purposes and may interfere with other functions in the pipeline.
```
Comment (pwittrock) [inline cmd/config/docs/api-conventions/functions-spec.md]: may interfere with other -> will likely impact subsequent

File: cmd/config/docs/api-conventions/functions-spec.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+A function SHOULD preserve comments when input serialization format is YAML. This allows for human authoring of configuration to coexist with changes made of functions.
```
Comment (pwittrock) [inline cmd/config/docs/api-conventions/functions-spec.md]: I suspect a large number of functions will not do this if written using libraries that don't preserve them.  Not sure if there is anything else to say on this though.

File: cmd/config/docs/api-conventions/functions-impl.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
+# Running Configuration Functions using kustomize CLI
+
+Configuration functions as defined [Configuration Functions Specification](functions-spec.md)
```
Comment (pwittrock) [inline cmd/config/docs/api-conventions/functions-impl.md]: links such as this one should probably include the command to pull up that document -- e.g. `kustomize help config docs-function-spec` or something

File: cmd/config/docs/api-conventions/functions-impl.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+can be implemented using any toolchain and invoked using any container workflow orchestrator
+like Tekton, Cloud Build, or run directly using `docker run`.
```
Comment (pwittrock) [inline cmd/config/docs/api-conventions/functions-impl.md]: like -> including

File: cmd/config/docs/api-conventions/functions-impl.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+Following is an example of running the `kustomize config run` using the preceding API.
+
+When run by `kustomize config run`, functions are run in containers with the
+following environment:
+
+- Network: `none`
+- User: `nobody`
+- Security Options: `no-new-privileges`
+- Volumes: the volume containing the `functionConfig` yaml is mounted under `/local` as `ro`
```
Comment (pwittrock) [inline cmd/config/docs/api-conventions/functions-impl.md]: Note: I think this was changed recently, we should update it

**Comment by pwittrock** [comment]
This looks good.  Should circulate it, and include the folks from kubebuilder

sig-am/kubebuilder: @DirectXMan12 @droot @justinsb 

sig-cli: @soltysh @monopole @seans3

**Comment by pwittrock** [comment]
@frankfarzan Would you send this to the sig-cli mailing list, and include it as an agenda item in the next sig-cli meeting?

File: cmd/config/docs/api-conventions/functions-spec.md (hunk 1 above)
Comment (frankfarzan) [reply cmd/config/docs/api-conventions/functions-spec.md]: Renamed to Use Cases.

File: cmd/config/docs/api-conventions/functions-impl.md (hunk 5 above)
Comment (frankfarzan) [reply cmd/config/docs/api-conventions/functions-impl.md]: Done.

File: cmd/config/docs/api-conventions/functions-impl.md (hunk 7 above)
Comment (frankfarzan) [reply cmd/config/docs/api-conventions/functions-impl.md]: How was it changed? Link to bug/commit?

File: cmd/config/docs/api-conventions/functions-spec.md (hunk 4 above)
Comment (pwittrock) [reply cmd/config/docs/api-conventions/functions-spec.md]: SGTM

File: cmd/config/docs/api-conventions/functions-impl.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
+# Running Configuration Functions using kustomize CLI
+
+Configuration functions can be implemented using any toolchain and invoked using any
+container workflow orchestrator including Tekton, Cloud Build, or run directly using `docker run`.
+
+Run `config help docs-fn-spec` to see the Configuration Functions Specification.
+
+`kustomize config run` is an example orchestrator for invoking Configuration Functions. This
+document describes how to implement and invike an example function.
```
Comment (marccampbell) [inline cmd/config/docs/api-conventions/functions-impl.md]: small typo:  ~invike~ invoke

Comment (frankfarzan) [reply cmd/config/docs/api-conventions/functions-impl.md]: Done.

**Pull request merged**